power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -130,4 +135,194 @@ inline void update_inverse(MainModelState<ComponentContainer> const& state, Forw
+namespace detail {
+
+constexpr Idx invalid_index{-1};
+
+template <typename T> bool check_id_na(T const& obj) {
+    if constexpr (requires { obj.id; }) {
+        return is_nan(obj.id);
+    } else if constexpr (requires { obj.get().id; }) {
+        return is_nan(obj.get().id);
+    } else {
+        throw UnreachableHit{"check_component_independence", "Only components with id are supported"};
+    }
+}
+
+struct UpdateCompProperties {
+    std::string name{};
+    bool has_any_elements{false};             // whether the component has any elements in the update data
+    bool ids_all_na{false};                   // whether all ids are all NA
+    bool ids_part_na{false};                  // whether some ids are NA but some are not
+    bool dense{false};                        // whether the component is dense
+    bool uniform{false};                      // whether the component is uniform
+    bool is_columnar{false};                  // whether the component is columnar
+    bool update_ids_match{false};             // whether the ids match
+    Idx elements_ps_in_update{invalid_index}; // count of elements for this component per scenario in update
+    Idx elements_in_base{invalid_index};      // count of elements for this component per scenario in input
+
+    constexpr bool no_id() const { return !has_any_elements || ids_all_na; }
+    constexpr bool qualify_for_optional_id() const {
+        return update_ids_match && ids_all_na && uniform && elements_ps_in_update == elements_in_base;
+    }
+    constexpr bool provided_ids_valid() const {
+        return is_empty_component() || (update_ids_match && !(ids_all_na || ids_part_na));
+    }
+    constexpr bool is_empty_component() const { return !has_any_elements; }
+    constexpr bool is_independent() const { return qualify_for_optional_id() || provided_ids_valid(); }
+    constexpr Idx get_n_elements() const {
+        assert(uniform || elements_ps_in_update == invalid_index);
+
+        return qualify_for_optional_id() ? elements_ps_in_update : na_Idx;
+    }
+};
+
+template <typename CompType> void process_buffer_span(auto const& all_spans, detail::UpdateCompProperties& properties) {
+    properties.ids_all_na = std::ranges::all_of(all_spans, [](auto const& vec) {
+        return std::ranges::all_of(vec, [](auto const& item) { return detail::check_id_na(item); });
+    });
+    properties.ids_part_na = std::ranges::any_of(all_spans,
+                                                 [](auto const& vec) {
+                                                     return std::ranges::any_of(vec, [](auto const& item) {
+                                                         return detail::check_id_na(item);
+                                                     });
+                                                 }) &&
+                             !properties.ids_all_na;
+
+    if (all_spans.empty()) {
+        properties.update_ids_match = true;
+    } else {
```

Comment:
From a previous comment on old PR, early return here please. :D

Reply:
Do you mean to put this early return at the very beginning of the function and return for empty input? Then we need to revisit and make sure `ids_all_na` and `ids_part_na` has good defaults.

If not, early return here is not needed.

Reply:
I am confused, the function is marked as `void`. What do you mean?

Reply:
```suggestion
    if (all_spans.empty()) {
        properties.update_ids_match = true;
        return;
    }
```
And remove bracket below.
And @Jerry-Jinfeng-Guo , you are right. Hence, early return at this place itself.